Adds CSR builder.

AndreLouisCaron opened the pull request:
This is a stripped down and rebased version of pull request #1914 with only the CSR builder.

AndreLouisCaron commented:
OK, guess I need to fix a few things to get this up to your standards :-)  I'll get around to that sometime tomorrow and update the PR.  I added basic docs for the new object.

Again, let me know if you want me to fix/change anything.

AndreLouisCaron commented:
Note that I'm having trouble figuring out which lines I missed in my tests because the Coveralls report says you need to authorize this repository.  You think that can be done?

reaperhulk commented:
If you click the coverage link coveralls posted you'll see what the PR coverage is... Unfortunately, coveralls does a *really* poor job of displaying it because it is very buggy. I'd suggest looking at your diff and trying to reason about what lines might have been missed :( Sorry we don't have a better answer for combined coverage right now...

AndreLouisCaron commented:
OK.  I'm still getting used to a few new (for me) things with this codebase... I'll keep working on this until I get something that gets a green from Travis.

AndreLouisCaron commented:
Any idea what all these errors are about?  Looks like Travis failed to create the virtual environments, but I've never seen this before...

    The command "./.travis/install.sh" failed and exited with 2 during .

alex commented:
test this please

AndreLouisCaron commented:
@alex I am testing...  Just added some more coverage!  However, it's a real pain to know where to add coverage since I only get a partial coverage when running locally...

alex commented:
@AndreLouisCaron Sorry for the confusion, that's a comment for jenkins to tell it to run the tests for your patch :-)

alex commented:
I definitely didn't mean to imply you weren't testing this.

On Wed, May 13, 2015 at 6:41 PM, André Caron <notifications@github.com>
wrote:

> @alex <https://github.com/alex> I am testing... Just added some more
> coverage! However, it's a real pain to know where to add coverage since I
> only get a partial coverage when running locally...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-101839561>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

AndreLouisCaron commented:
@alex Ah!  I see :-)  No worries!

AndreLouisCaron commented:
OK, I got almost everything covered.  The remaining code coverage is related to the `CertificateSigningRequest.extensions` property.  I copied the basic structure from `Certificate.extensions`: things like duplicate extensions, unsupported extensions and unsupported critical extensions.  The tests for that implementation use pre-existing files in the `cryptography_vectors` package.

I would need the following files (based on the similar vectors for the `Certificate.extensions` property tests):»
- `two_basic_constraints_csr.pem`;
- `unsupported_extension_csr.pem`; and
-  `unsupported_extension_critical_csr.pem`.

Since the documentation refers to this package as a set of trusted vectors, do you prefer adding them yourselves or should I generate them?

Thanks!

AndreLouisCaron commented:
I added some test vectors to see if I can (finally) reach 100% coverage.  Feel free to replace them with your own if you see fit.

AndreLouisCaron commented:
BTW, even if this reaches a green build, I'd like to squash all these fixup commits before you merge it.

reaperhulk commented:
jenkins, add to whitelist

AndreLouisCaron commented:
Looks like I finally managed to get a green build :-)  Let me know if you want me to make any touch ups to this PR.

reaperhulk commented:
I look forward to reviewing this (although I'm sure @alex will beat me to it), but now that you've got it working would you mind splitting it into multiple PRs? PRs over ~300 lines become very difficult to effectively review. At a glance it looks like there are a few things you could slice out into their own PRs to lower the line count here:

* C bindings
* Refactoring _build_basic_constraints to a top level function
* The new vectors (which will require an entry in test-vectors.rst to describe them BTW)
* `public_bytes` (...do we really not have serialization for CSR or certificates still? Did I only implement that for public/private keys?)

Thanks for your work so far!

reaperhulk commented:
(Sorry for the additional brain dump):

It looks like we're going to need to refactor every `_build_*` method into a function so they can be shared across Certificate and CSR. Maybe the PR should just do that for all of them at once?

We can land many of these PRs pretty quickly so we can focus on the more complex CSRBuilder :)

AndreLouisCaron commented:
I don't mind splitting this PR, but I'm only working on this part time and I won't be able to do that until the weekend.

For the `_build_*` methods, I'll do a single PR.  Do you mind if I rename them to `_decode_*()` since we're going to need the reverse `_encode_*()` methods?  I named the first one `_make_basic_constraints()` but I'm not happy with the name.  Feel free to suggest something else.

For the `public_bytes()` method, I didn't find any existing way to do it.  I'm not super familiar with all of this  codebase, so maybe it already exists.

Thanks for being supportive, it's really appreciated.  Hopefully, now that I'm getting the hang of this, the next contribution will be faster/smoother!

reaperhulk commented:
No problem on the timing. We just released 0.9 last night so this feature will make it into the next major release (we aim to do those sorts of releases every 4-8 weeks).

decode sounds reasonable to me for the conversion from OpenSSL internal repr to our python classes and then encode is a decent name for taking a python version and building the openssl structs.

It doesn't appear that we support serializing to bytes on certificate or CSR. That's probably because we didn't have any way to create these objects except by loading the PEM/DER until you came along. I'm not 100% sure we want to call it `public_bytes` (although as a counter-argument, everything in a CSR/Cert is public data so maybe we should) but we can discuss that in the PR.

alex commented:
I split the bindings out here, so if you rebase on master that'll be taken care of :-)

AndreLouisCaron commented:
OK, quick recap:

- @alex Added the bindings in PR #1955 (thanks!)
- I moved the `_build_*()` method refactoring to PR #1961 (all of them at once)
- I moved the new CSR test vectors to PR #1962
- I'll move support for CSR extensions in a new PR (to be based on #1962)
- I moved the CSR writing support to PR #1963

When all of the above are merged, I'll strip this PR down to only the CSR builder, its documentation and its tests.

AndreLouisCaron commented:
Looks like I messed up my rebase.  Lemme fix that up.

AndreLouisCaron commented:
Travis build failed because of a network timeout contacting `pypi.python.org`.  Let me know if you see anything I should fix.

alex commented:
@AndreLouisCaron restarted the build, thanks

alex commented:
This needs a rebase due to merge conflicts. Thanks!

AndreLouisCaron commented:
@alex Rebase complete!

AndreLouisCaron commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,88 @@
+    # Allocate a buffer for encoded payload.
+    cdata = backend._ffi.new(
+        'unsigned char[]',
+        2048,  # TODO: shrink to fit!
```
Note to reviewers: I have no idea how many bytes to allocate here.  If anyone has a better suggestion than this (probably too large) wild guess, I'm all ears!

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,88 @@
+    p = backend._ffi.new('unsigned char*[1]')
+    assert p != backend._ffi.NULL
+    p[0] = cdata
+    r = backend._lib.i2d_BASIC_CONSTRAINTS(constraints, p)
```
i2d_BASIC_CONSTRAINTS will allocate its own unsigned char * and return it in the unsigned char ** so there's no need to do the cdata allocation. Instead you can just do

```python
pp = backend._ffi.new("unsigned char **")
res = backend._lib.i2d_BASIC_CONSTRAINTS(constraints, pp)
```
and now pp[0] is the unsigned char * you need.

You will need to free that memory though:

```python
    pp = backend._ffi.gc(
        pp, lambda pp: backend._lib.OPENSSL_free(pp[0])
    )
```

AndreLouisCaron replied:
Thanks, works like a charm!

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,80 @@
+
+
+def _encode_name(backend, attributes):
+    resolve = {
```
We use `OBJ_obj2txt` to grab an ASN1 object and turn it into a textual OID in backends/openssl/x509.py. We should be able to use `OBJ_txt2obj` to convert an OID back into an OBJ and then use `X509_NAME_add_entry_by_OBJ` rather than building this incomplete mapping. (`X509_NAME_add_entry_by_OBJ` will, of course, need to be added as a binding).

reaperhulk replied:
Still an issue, just dropping another comment on it because on larger PRs it's easy to not notice these sometimes :)

AndreLouisCaron replied:
BTW, this still isn't fixed.  On it now :-)

AndreLouisCaron replied:
Created PR #2000 to add the missing binding.  Will update this PR when the binding is merged.

AndreLouisCaron replied:
This issue is now fixed!

reaperhulk commented:
I'm not a big fan of the `set_*` methods. What do people think about making these just setter properties? e.g.

```python
builder = CertificateRequestBuilder()
builder.version = x509.Version.v3
builder.version = 3 # raises error
```

We could potentially do the same thing with extensions (make it an empty list and .append onto it), although that might be confusing since you wouldn't be able to validate what was added was valid until you called sign.

@alex do you have an opinion?

etrauschke commented:
In general I'm not a big fan of the way this is supposed to work. I mainly have an issue with backend.sign_x509_request().
This creates a new x509_REQ object and signs it, so the name is misleading. I furthermore would think that the work flow for this should be different.

I'd think we would have a sign() method in the CertificateSigningRequest class which does the signing. Then the builder would just create an instance of  CertificateSigningRequest which someone then can run sign() on. Or better, just create a CertBuilder which will take a CSR object as an optional argument to __init__() to set all the properties for the cert.

AndreLouisCaron commented:
@etrauschke I'm definitely open to changes in the interface, but I'm not comfortable with introducing the concept of "signed" and "unsigned" certificate requests.  Currently, the `CertificateSigningRequest` interface is immutable and it is always signed.  I'm not sure I want to break that.

It also seems to me like the whole [builder pattern](http://en.wikipedia.org/wiki/Builder_pattern)'s purpose is to provide a full-fledged object.

I'm also unsure about your cert builder idea.  In the typical flow, the client signs their certificate request and sends it to the CA, then the CA verifies it, extracts the subject's info and requested extensions and then builds a certificate -- all the while filtering, transforming and adding new extensions according to its own policy.  I don't think you can bypass the CSR signing step by the client nor assume that the CA will pass the request as is to a new certificate.

Maybe it's just the name that's misleading?  Anybody have any thoughts on this?

alex commented:
First, I agree with @AndreLouisCaron that a CSR object should always be "complete" (meaning signed).

In terms of the API, I'd like for builders to be immutable and use a chaining pattern:

```python

CertificateSigningRequestBuilder().version(v3).subject(...).extension(SAN()).sign(private_key, backend)
```

The naming probably needs some work, but this feels like a good pattern to me.

AndreLouisCaron commented:
@alex As we already discussed, I'll add the method chaining as soon as I know the rest of the interface is rather stable.  If we can figure out the names and the flow first, I'd be really grateful :-)

AndreLouisCaron commented:
@reaperhulk I don't mind using properties and I wouldn't mind removing the `set_` prefix to setters (as Alex suggested) either.

I also haven't added any getters, which might be desirable -- especially in an interactive tool to build CSRs.  Let me know if you want me to add those inline or in a separate PR.

etrauschke commented:
One thing to keep in mind, the work flow of CSR -> forward to external CA
-> receive cert from CA is not the only one.
This might be the case for someone who creates a cert for a server but when
using x509 objects programmatically this is often not the case.

For some use cases the Python app generating the x509 objects is catually
in control of the CA. I have one of these use cases and we are generating
client certs from our own CA. So we don't need to create CSRs, we can
create certificates directly and sign them.

I can live with the principle of having cert objects read-only and the need
to have factories to build them, but don't assume that a CSR is required to
get a valid certificate. The CSR is just a method to send requests to an
external CA. And if you want an actual CA to use cryptography to create
certificates you might want to offer a way for them to easily create a cert
from a given CSR, without having to copy things like issuer, subject,
extensions, etc. manually.

Erik

On Tue, May 19, 2015 at 3:49 PM, André Caron <notifications@github.com>
wrote:

> @etrauschke <https://github.com/etrauschke> I'm definitely open to
> changes in the interface, but I'm not comfortable with introducing the
> concept of "signed" and "unsigned" certificate requests. Currently, the
> CertificateSigningRequest interface is immutable and it is always signed.
> I'm not sure I want to break that.
>
> It also seems to me like the whole builder pattern
> <http://en.wikipedia.org/wiki/Builder_pattern>'s purpose is to provide a
> full-fledged object.
>
> I'm also unsure about your cert builder idea. In the typical flow, the
> client signs their certificate request and sends it to the CA, then the CA
> verifies it, extracts the subject's info and requested extensions and then
> builds a certificate -- all the while filtering, transforming and adding
> new extensions according to its own policy. I don't think you can bypass
> the CSR signing step by the client nor assume that the CA will pass the
> request as is to a new certificate.
>
> Maybe it's just the name that's misleading? Anybody have any thoughts on
> this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-103689764>.
>

AndreLouisCaron commented:
@etrauschke Indeed, that's a totally valid use case.  That's why the `CertificateBuilder` in PR #1980 doesn't work with CSRs...

etrauschke commented:
Ok, so what is the expected work flow when someone who wants to write a CA
application based on cryptography? Read the CSR from the PEM file, unpack
all the properties then start the CertificateBuilder and run all the set_*
functions? Seems painful to me and prone to missing stuff.

Erik

On Wed, May 20, 2015 at 3:21 PM, André Caron <notifications@github.com>
wrote:

> @etrauschke <https://github.com/etrauschke> Indeed, that's a totally
> valid use case. That's why the CertificateBuilder in PR #1980
> <https://github.com/pyca/cryptography/pull/1980> doesn't work with CSRs...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-104060903>.
>

etrauschke commented:
One more comment about the general approach. So far it seemed like we modeled the classes in cryptography very closely to the associated x509 objects. I think it could be useful to extend this to certificate, csr and crl objects as well. I still think the factories are useful to create certs but also like the idea of actually being able to manipulate existing certs and then store them into pem files. Anyway, this could also be added later on.

AndreLouisCaron commented:
@etrauschke Yeah, that's the idea with the builders.  However, I consider this to be a minimalist low-level interface (as per the "hazmat" layer) and I expect that we can build some higher-level recipes out of this interface.

While I agree with the fact that we should be able to "edit" a certificate, I like the fact that the certificate is an immutable object -- IRL, it *is* immutable once it's signed.  I consider this use case to be one where you open up a cert, create a new builder, feed it with what you already have minus what you want to remove, add what you want to add and then sign the result to create a new certificate.  Then, you can simply save it to the same PEM file you loaded the original one from if you like.  Again, some higher-level recipes might help with implementing this quickly and correctly.

alex commented:
IMO the way we should provide to do the CA-like CSR -> Cert is a specific
function for that purpose, which includes a policy (which defaults to the
RFC XXX defaults) on how to trust the info from the CSR (e.g. use these
extensions, drop all others, etc.).

On Wed, May 20, 2015 at 6:38 PM, André Caron <notifications@github.com>
wrote:

> @etrauschke <https://github.com/etrauschke> Yeah, that's the idea with
> the builders. However, I consider this to be a minimalist low-level
> interface (as per the "hazmat" layer) and I expect that we can build some
> higher-level recipes out of this interface.
>
> While I agree with the fact that we should be able to "edit" a
> certificate, I like the fact that the certificate is an immutable object --
> IRL, it *is* immutable once it's signed. I consider this use case to be
> one where you open up a cert, create a new builder, feed it with what you
> already have minus what you want to remove, add what you want to add and
> then sign the result to create a new certificate. Then, you can simply save
> it to the same PEM file you loaded the original one from if you like.
> Again, some higher-level recipes might help with implementing this quickly
> and correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-104063764>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

AndreLouisCaron commented:
@alex Fair enough.  I guess that should be in a separate PR though?

alex commented:
Yup :-)

On Wed, May 20, 2015 at 7:01 PM, André Caron <notifications@github.com>
wrote:

> @alex <https://github.com/alex> Fair enough. I guess that should be in a
> separate PR though?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-104067157>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

AndreLouisCaron commented:
@alex @reaperhulk Ping?  Should I go ahead and switch to properties?  Should I add anything else?  Do I go ahead and add method chaining?

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,80 @@
 
+def _encode_asn1_int(backend, x):
+    i = backend._lib.ASN1_INTEGER_new()
+    # i = backend._ffi.gc(i, backend._lib.ASN1_INTEGER_free)
```
This should be removed and a comment should be added at the top of this function stating that the generated object will not be garbage collected (to help remind us that we need to either know it's being added to something that will own it or that we need to manually register for gc).

AndreLouisCaron replied:
Copied and adapted the a similar comment from `Backend._int_to_bn()`.

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,80 @@
+
+def _encode_asn1_str(backend, x, n):
+    s = backend._lib.ASN1_OCTET_STRING_new()
+    # s = backend._ffi.gc(s, backend._lib.ASN1_OCTET_STRING_free)
```
(Same)

AndreLouisCaron replied:
Took a look at the source code for this one.  It seems like `X509_EXTENSION_create_by_OBJ()` passes the data to `X509_EXTENSION_set_data()` which creates a copy of the data without taking ownership.  I uncommented the GC line and everything passes without crashing, so I'll submit it with the GC statement to avoid a leak.

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,80 @@
+        name = name.encode('ascii')
+    obj = backend._lib.OBJ_txt2obj(name, 1)
+    assert obj != backend._ffi.NULL
+    # NOTE: not sure if we should GC...
```
Looks like this is still outstanding. Based on your calls you probably need to look at the implementation of `X509_EXTENSION_create_by_OBJ` to determine if it frees/retains the ASN1_OBJECT you pass to it?

AndreLouisCaron replied:
Took a look at the source code for this one. It seems like `X509_EXTENSION_create_by_OBJ()` passes the data to `X509_EXTENSION_set_obj()` which creates a copy of the ASN.1 object without taking ownership. I added the GC code and everything passes without crashing, so I'll submit it with the GC statement to avoid a leak.

reaperhulk commented:
@AndreLouisCaron Yeah I think you should move to properties and do the method chaining (seems like that'd be difficult to put in a separate PR without changing the API significantly). There are also a few outstanding comments to address.

etrauschke commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +760,67 @@ def cmac_algorithm_supported(self, algorithm):
         return _CMACContext(self, algorithm)
 
+    def sign_x509_request(self, builder, private_key, algorithm):
+        # TODO: check type of private key parameter.
```
We decided to keep the general mechanism but this name still doesn't seem appropriate. What you do is create a CSR, not just sign it, so I'd call it create_x509_request().

AndreLouisCaron replied:
Renamed to `create_x509_csr()` and added this to the X.509 backend interface too.

etrauschke commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +760,67 @@ def cmac_algorithm_supported(self, algorithm):
+            x509_req, private_key._evp_pkey, evp_md
+        )
+        assert res > 0
+
```
If you wanna be anal here you can compare if res equals the length of the key you are signing with.

AndreLouisCaron replied:
Sure, I don't mind adding that.  However, I'm not sure how to obtain the desired length (seems like the length depends on the size of the selected hash.  Any ideas?

Also, I guess I'll need to add a test for the case where this does _not_ come back equal.  Any suggestion for a way to test this?

reaperhulk commented:
Bumping to mention this will need a rebase now

AndreLouisCaron commented:
Rebased and fixed some of the issues.  Getting on method chaining now :-)

I've been thinking about this and I don't think I'll be able to use property assignments _and_ method chaining -- even if Python supported statements as expressions, it would probably give an ugly result (think `(r.version = v3).subject = "foo"`).  Do you prefer method chaining or property assignments?

Also, do you have a reference style to implement method chaining?  I guess I can just copy all the builder's data members, but maybe you have something else in mind?

alex commented:
Method chaining please, it's the only way to do immutability.

On Sun, May 31, 2015 at 5:07 PM, André Caron <notifications@github.com>
wrote:

> Rebased and fixed some of the issues. Getting on method chaining now :-)
>
> I've been thinking about this and I don't think I'll be able to use
> property assignments *and* method chaining -- even if Python supported
> statements as expressions, it would probably give an ugly result (think (r.version
> = v3).subject = "foo"). Do you prefer method chaining or property
> assignments?
>
> Also, do you have a reference style to implement method chaining? I guess
> I can just copy all the builder's data members, but maybe you have
> something else in mind?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927#issuecomment-107246551>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

AndreLouisCaron commented:
I believe I've addressed all issues and the implementation now supports method chaining.  Let me know if you want me to change anything else!

reaperhulk commented on src/cryptography/hazmat/backends/interfaces.py:
```diff
@@ -274,6 +274,12 @@ def load_pem_x509_csr(self, data):
+    @abc.abstractmethod
+    def create_x509_csr(self, builder, private_key, algorithm):
+        """
+        Create and sign an X.509 CSR from a CSR buidler object.
```
typo on builder

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,86 @@
+    # Fetch the encoded payload.
+    pp = backend._ffi.new('unsigned char**')
+    assert pp != backend._ffi.NULL
+    r = backend._lib.i2d_BASIC_CONSTRAINTS(constraints, pp)
```
I believe the dereferenced value of `pp` will leak here unless you register pp[0] against the gc with `OPENSSL_free` after pp falls out of scope?

AndreLouisCaron replied:
Seems reasonable.  I took a look at the OpenSSL source, but I can't seem to wrap my head around the macros used in the `i2d_*()` functions.  I added a call to `OPENSSL_free(pp[0])` and the tests pass without crashing.

reaperhulk commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,86 @@
+
+    """
+    i = backend._lib.ASN1_INTEGER_new()
+    backend._lib.ASN1_INTEGER_set(i, x)
```
This call will fail for any integer > native word size. We should probably convert the python int to BN and then do a conversion of BN to ASN1_INTEGER (I assume a function already exists for this purpose) so we can safely use this as a general purpose converter.

AndreLouisCaron replied:
Fixed!

reaperhulk commented on src/cryptography/x509.py:
```diff
@@ -1344,3 +1344,52 @@ def extensions(self):
+        self._subject_name = subject_name
+        self._extensions = extensions[:]
+
+    def set_version(self, version):
```
Unlike certificates, version 1 is the only version of x509 CSRs that is officially defined. We raise errors on parsing certificates that have versions not defined by the RFC so to be consistent we should probably just remove `set_version` and force it to version 1.

That said, looking at our vectors we have quite a few custom CSRs that are defined as 0x2. That is likely an artifact of the way I generated them? Bears further investigation.

AndreLouisCaron replied:
Actually, the ones I added recently are all generated with version 3 (0x2).  If the code that loads CSRs is supposed to raise when parsing "v3" certificates, the code is probably broken because those CSRs load fine in the test suite.

I'll let you check whether that code should be fixed.  Should I go an fix the test vectors that are incorrectly set to v3 (and then remove the `set_version()` method from the builder)?

reaperhulk replied:
We don't parse CSR version at all right now which is how that is being ignored. I'm not 100% sure what we should do here, but to my knowledge there's no actual 0x2 version for CSRs. RFC 2986 defines v1.7 of PKCS#10 (unless something obsoletes it, which I have been unable to find) and only defines a v1. In practice since it's an ASN.1 integer you can obviously put whatever you want there.

AndreLouisCaron replied:
OK.  I'll send in a separate PR to fix the CSR test vectors.  I'll also remove the `set_version()` function since it seems like it can only be misused at the moment.

AndreLouisCaron replied:
Added a commit which removes the `set_version()` method.

reaperhulk commented on docs/x509.rst:
```diff
@@ -468,6 +468,50 @@ X.509 Revoked Certificate Object
+X.509 CSR (Certificate Signing Request) Builder Object
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. class:: CertificateSigningRequestBuilder
```
This should have a versionadded and also a doctest showing it in action :)

reaperhulk commented on docs/x509.rst:
```diff
@@ -468,6 +468,50 @@ X.509 Revoked Certificate Object
+
+        Creates an empty certificate signing request.
+
+    .. method:: set_version(version)
```
No longer exists.

reaperhulk commented:
A few new nits, but from the perspective of the OpenSSL backend code this looks correct now. I'd like @alex to take a closer look at the API but the substance of it looks good to me now.

Do we need to land CSR extension parsing before we land this so we can parse what we're generating?

AndreLouisCaron commented:
OK, I'll address the remaining issues tomorrow (unless Alex finds more :p)!

> Do we need to land CSR extension parsing before we land this so we can parse what we're generating?

You already merged basic CSR extension support in PR #1966.  I will probably be needing some more extensions in in the future and I'll probably add them along the way.  What did you have in mind?

reaperhulk commented:
Whoops forgot I had done that :)



> On Jun 1, 2015, at 11:14 PM, André Caron <notifications@github.com> wrote:
> 
> OK, I'll address the remaining issues tomorrow (unless Alex finds more :p)!
> 
> Do we need to land CSR extension parsing before we land this so we can parse what we're generating?
> 
> You already merged basic CSR extension support in PR #1966. I will probably be needing some more extensions in in the future and I'll probably add them along the way. What did you have in mind?
> 
> —
> Reply to this email directly or view it on GitHub.
>

AndreLouisCaron commented:
The build failure in the OSX worker looks like some kind of timeout.  Can you restart the Travis build?

reaperhulk commented:
A reasonably common problem. I've restarted the build.

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
+
+        Creates an empty certificate signing request.
+
+    .. method:: set_subject_name(name)
```
@reaperhulk do you have an opinion on naming here; I'm thinking we drop teh `set_` prefix?

reaperhulk replied:
I would prefer "set" not being present, yes.

AndreLouisCaron replied:
I don't mind changing this at all, but I'm wondering what the interface for eventual getter methods/properties would have..?  If we rename `set_subject_name()` to `subject_name()`, then how would you call the getter?

alex replied:
I'm of the opinion that Builders shouldn't have any getters.

On Sat, Jun 6, 2015 at 6:36 PM, André Caron <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1927#discussion_r31872134>:
>
> > +        >>> buidlder = builder.add_extension(x509.Extension(
> > +        ...     x509.OID_BASIC_CONSTRAINTS,
> > +        ...     True,
> > +        ...     x509.BasicConstraints(False, None),
> > +        ... ))
> > +        >>> request = builder.sign(
> > +        ...     default_backend(), private_key, hashes.SHA1()
> > +        ... )
> > +        >>> isinstance(request, x509.CertificateSigningRequest)
> > +        True
> > +
> > +    .. method:: __init__()
> > +
> > +        Creates an empty certificate signing request.
> > +
> > +    .. method:: set_subject_name(name)
>
> I don't mind changing this at all, but I'm wondering what the interface
> for eventual getter methods/properties would have..? If we rename
> set_subject_name() to subject_name(), then how would you call the getter?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1927/files#r31872134>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

AndreLouisCaron replied:
Method renamed to `subject_name()`.

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
+    .. method:: set_subject_name(name)
+
+        :param name: The :class:`Name` of the certificate subject.
+        :returns: A new `CertificateSigningRequestBuilder`.
```
Please use the full `:class:` prefix so everything linkifies. Sorry, I know it's a ton of extra typing :-/

AndreLouisCaron replied:
Fully qualified all symbol references in the CSR builder class.

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
+        :param name: The :class:`Name` of the certificate subject.
+        :returns: A new `CertificateSigningRequestBuilder`.
+
+    .. method:: add_extension(extension)
```
@reaperhulk what do you think about making the API `add_extension(BasicConstraints(...), critical=True)`, and have it auto-infer the OID?

reaperhulk replied:
That sounds reasonable. It would preclude this API being used to do custom extensions, but that's not possible right now anyway since the underlying backend needs to know how to serialize the extension (and we have no plans to allow users to define arbitrary serializers).

AndreLouisCaron replied:
Changed the API to infer the extension OID.

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,73 @@ X.509 Revoked Certificate Object
+
+        :param algorithm: The
+            :class:`~cryptography.hazmat.primitives.hashes.HashAlgorithm`
+            that will be used to generate the request signature.
```
@reaperhulk Does this full express the set of ways we might want to sign it?

reaperhulk replied:
For x509 it covers it. You don't get to pick your padding in x509 (thankfully). That said, the docs should list RSA/DSA/EC keys, not just RSA for `private_key`.

AndreLouisCaron replied:
Added reference to DSA and EC private keys in the docs.

alex commented:
There's also a merge conflict here :-(

AndreLouisCaron commented:
Addressed all new issues and rebased to fix the merge conflict :-)

alex commented:
Will aim to give this a thorough review this evening. Thanks for driving this forwards! :sparkles:

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,80 @@ X.509 Revoked Certificate Object
+        ...     x509.NameAttribute(x509.OID_COMMON_NAME, 'cryptography.io'),
+        ... ]))
+        >>> buidlder = builder.add_extension(
+        ...     x509.BasicConstraints(False, None), critical=True,
```
Can we use keyword arguments in the `BasicConstraints` constructor.

sigmavirus24 replied:
Fixed in 9d90171

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,80 @@ X.509 Revoked Certificate Object
+        ...     x509.BasicConstraints(False, None), critical=True,
+        ... )
+        >>> request = builder.sign(
+        ...     default_backend(), private_key, hashes.SHA1()
```
We should use `SHA256` for the example.

sigmavirus24 replied:
Fixed in 9d90171

alex commented on docs/x509.rst:
```diff
@@ -468,6 +468,80 @@ X.509 Revoked Certificate Object
+        >>> isinstance(request, x509.CertificateSigningRequest)
+        True
+
+    .. method:: __init__()
```
I think we can drop the `__init__` docs, it's pretty self evidence.

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+    return i
+
+
+def _encode_asn1_str(backend, x, n):
```
Can we use slightly more descriptive variable names here, perhaps "data" and "length"

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+    for attribute in attributes:
+        value = attribute.value
+        if isinstance(value, six.text_type):
+            value = value.encode('ascii')
```
Don't we require values to be bytes in the constructor? When is this path hit?

alex replied:
I got this backwards, we require values to be `unicode`, so we can drop the `if`

sigmavirus24 replied:
Done in https://github.com/sigmavirus24/cryptography/commit/eb1549ddd970b78d1fdddfe8ad2c7b60a88bede0 but I'm not sure that using `'ascii'` is right in this case.

reaperhulk replied:
Yeah it's possible we should be using `'utf8'`. Additionally we maybe shouldn't use `MBSTRING_ASC` here. `MBSTRING_UTF8` may be more appropriate. This is going to require some investigation.

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+    ASN1_OBJECT.
+    """
+    if isinstance(name, six.text_type):
+        name = name.encode('ascii')
```
Same question here.

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+
+    # Fetch the encoded payload.
+    pp = backend._ffi.new('unsigned char**')
+    assert pp != backend._ffi.NULL
```
You can drop this assert, `new` never returns NULL

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+    return obj
+
+
+def _encode_basic_constraints(backend, ca=False, pathlen=0, critical=False):
```
The defaults can be dropped, they're never used.

alex replied:
In fact, this should probably take a `BasicConstraints` instance as an argument.

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+        constraints.pathlen = _encode_asn1_int(backend, pathlen)
+
+    # Fetch the encoded payload.
+    pp = backend._ffi.new('unsigned char**')
```
Space after "char"

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,99 @@
+    assert extension != backend._ffi.NULL
+
+    # Release acquired memory.
+    backend._lib.OPENSSL_free(pp[0])
```
This should be done using `ffi.gc()`

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +779,67 @@ def cmac_algorithm_supported(self, algorithm):
+        assert evp_md != self._ffi.NULL
+
+        # Create an empty request.
+        x509_req = self._lib.X509_REQ_new()
```
This needs a free-function added with `ffi.gc`, right now the memory leaks as far as I can tell.

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +779,67 @@ def cmac_algorithm_supported(self, algorithm):
+                    extension.critical
+                )
+            else:
+                raise ValueError('Extension not yet supported.')
```
This should be `NotIMplementedError` I guess

sigmavirus24 replied:
Fixed in 9d90171

alex commented on src/cryptography/x509.py:
```diff
@@ -1344,3 +1344,41 @@ def extensions(self):
+        """
+        if not isinstance(name, Name):
+            raise TypeError('Expecting x509.Name object.')
+        return CertificateSigningRequestBuilder(name, self._extensions)
```
Should it be an error to call `subject_name` twice?

reaperhulk replied:
I have no opinion here. @alex did you lean one way?

alex replied:
I think I'm leaning towards "yes" on an error -- it's easier to remove the error if we find that it would be useful than it would be to add it in later

sigmavirus24 replied:
See sigmavirus24@5b33525

alex commented on tests/test_x509.py:
```diff
@@ -644,6 +644,127 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
+                x509.NameAttribute(x509.OID_COMMON_NAME, 'cryptography.io'),
+            ])
+        ).add_extension(
+            x509.BasicConstraints(True, 2), critical=True
```
pass the args to `BasicConstraints` as keywords please.

sigmavirus24 replied:
Fixed in 9d90171

alex replied:
It doesn't appear to have happened here.

sigmavirus24 replied:
You're right. I thought I had updated all of them but it appears I missed one.

sigmavirus24 replied:
Actually fixed in https://github.com/sigmavirus24/cryptography/commit/eece6d4ce8e50ed0f4869b062a5c77679e797b34

alex commented on tests/test_x509.py:
```diff
@@ -644,6 +644,127 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
+                x509.NameAttribute(x509.OID_COMMON_NAME, 'cryptography.io'),
+            ])
+        ).add_extension(
+            x509.BasicConstraints(False, None), critical=True,
```
Same here

sigmavirus24 replied:
Fixed in 9d90171

sigmavirus24 commented:
@AndreLouisCaron hopefully 9d90171 helps move this along. You'll also need to rebase this again.

AndreLouisCaron commented:
I've been busy with some other things for a short while.  I should be able to finish this this week.  Thanks to @sigmavirus24 for pushing some of the final fixes for the code review!

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -56,6 +56,100 @@
+
+    pp[0] = backend._ffi.gc(pp[0], backend._lib.OPENSSL_free)
+    # Release acquired memory.
+    pp[0] = backend._ffi.NULL
```
These 3 lines should be removed, and under `assert r > 0` we should have something more like: https://github.com/pyca/cryptography/blob/master/src/cryptography/hazmat/backends/openssl/x509.py#L40-L47

sigmavirus24 replied:
@alex is https://github.com/sigmavirus24/cryptography/commit/bac0adf867fc01355e0ec80335d82b829f6c671b correct?

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +780,67 @@ def cmac_algorithm_supported(self, algorithm):
+        # Create an empty request.
+        x509_req = self._lib.X509_REQ_new()
+        x509_req = self._ffi.gc(x509_req, self._lib.X509_REQ_free)
+        assert x509_req != self._ffi.NULL
```
Whoops, this assert should go above the `gc` line

sigmavirus24 replied:
See https://github.com/sigmavirus24/cryptography/commit/eece6d4ce8e50ed0f4869b062a5c77679e797b34

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +780,67 @@ def cmac_algorithm_supported(self, algorithm):
+        public_key = private_key.public_key()
+        res = self._lib.X509_REQ_set_pubkey(
+            x509_req, public_key._evp_pkey
+        )
```
@reaperhulk does this get the refcounting right?

reaperhulk replied:
Yes, it looks like it encodes its own copy when calling X509_REQ_set_pubkey (which calls X509_PUBKEY_set), so this is correct.

alex commented on src/cryptography/hazmat/backends/openssl/backend.py:
```diff
@@ -686,6 +780,67 @@ def cmac_algorithm_supported(self, algorithm):
+        assert res == 1
+
+        # Add extensions.
+        extensions = self._lib.sk_X509_EXTENSION_new_null()
```
There should be an `assert extesion != ffi.NULL`

reaperhulk replied:
I assume you mean there should be an assert after this call?

sigmavirus24 replied:
I took care of that in my branch. (Probably should have commented like I did with the last round of reviews.)

sigmavirus24 replied:
See https://github.com/sigmavirus24/cryptography/commit/eece6d4ce8e50ed0f4869b062a5c77679e797b34

alex commented on src/cryptography/x509.py:
```diff
@@ -1344,3 +1344,41 @@ def extensions(self):
+        Creates an empty X.509 certificate request (v1).
+        """
+        self._subject_name = subject_name
+        self._extensions = extensions[:]
```
I don't think a copy is necessary here.

sigmavirus24 replied:
Fixed in https://github.com/sigmavirus24/cryptography/commit/eece6d4ce8e50ed0f4869b062a5c77679e797b34

alex commented on tests/test_x509.py:
```diff
@@ -644,6 +644,127 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
+            public_exponent=65537,
+            key_size=2048,
+            backend=backend,
+        )
```
Instead of generating a key, we should use onf the fixtures.

sigmavirus24 replied:
See https://github.com/sigmavirus24/cryptography/commit/890c749a8ea210b7028d03c1d0d4c3bc27c93b54

alex commented on tests/test_x509.py:
```diff
@@ -644,6 +644,127 @@ def test_public_bytes_match(self, request_path, loader_func, encoding,
+            public_exponent=65537,
+            key_size=2048,
+            backend=backend,
+        )
```
Same

sigmavirus24 replied:
See https://github.com/sigmavirus24/cryptography/commit/890c749a8ea210b7028d03c1d0d4c3bc27c93b54

alex commented:
We're getting very close! Thanks so much for sticking with this.

It also needs a rebase.

sigmavirus24 commented:
@AndreLouisCaron I've tried to address a bunch of the new review comments in https://github.com/sigmavirus24/cryptography/tree/csr-builder and I've rebased the branch.

I'm still not sure what the right answers are for:

- https://github.com/pyca/cryptography/pull/1927#discussion_r31873303
- https://github.com/pyca/cryptography/pull/1927#discussion_r32617658

So someone else may have to address those.

FWIW, I'm deliberately not "taking over" this pull request because there's a history of this feature here and I'd like to not lose that by making a new pull request for this. I'm just hoping to help @AndreLouisCaron wrap this up.

reaperhulk commented:
closing fixed, this was merged as #2045
